File: Javascript/Sorting/Merge_Sort.js



// Merge Sort in javascript

function Merge_Sort(arr,l,h){

    if(l>=h){
      return
    }
    let mid=Math.floor((l+h)/2);
    Merge_Sort(arr,l,mid);
    Merge_Sort(arr,mid+1,h);
    Merge(arr,l,mid,h);
    
  }
  
  function Merge(arr,l,mid,h){
    let temp=[];
    let left=l;
    let right=mid+1;
    while(left<=mid && right<=h){
      if(arr[left]<=arr[right]){
        temp.push(arr[left])
        left++;
      }
      else{
        temp.push(arr[right]);
        right++;
      }
    }
    while(left<=mid){
      temp.push(arr[left]);
      left++;
    }
    while(right<=h){
      temp.push(arr[right]);
      right++;
    }
    
    for(let i=l;i<=h;i++){
      arr[i]=temp[i-l];
    }
  }
  
  
  function main(){
    let arr=[9,2,6,1,28,3,5];
    Merge_Sort(arr,0,arr.length-1);
    console.log(arr)
  }
  
  main()

File: C++/Sorting/Merge_Sort.cpp
#include <iostream>
#include <bits/stdc++.h>
using namespace std;

void Merge(vector<int> &arr,int low,int mid,int high){
  vector<int>temp;
  int left=low;
  int right=mid+1;
  
  
  while(left<=mid && right<=high){
    if(arr[left]<=arr[right]){
      temp.push_back(arr[left]);
      left++;
    }
    else{
      temp.push_back(arr[right]);
      right++;
      
    }
  }
  while(left<=mid){
    temp.push_back(arr[left]);
    left++;
  }
  while(right<=high){
    temp.push_back(arr[right]);
    right++;
  }
  
  for(int i=low;i<=high;i++){
    arr[i]=temp[i-low];
  }
}


void MergeSort(vector<int> & arr,int low,int high){
  if(low>=high)return;
  int mid=(low+high)/2;
  MergeSort(arr,low,mid);
  MergeSort(arr,mid+1,high);
  Merge(arr,low,mid,high);
}


int main() 
{
    vector<int> arr = {9, 4, 7, 6, 3, 1, 5}  ;
    int n = 7;
    MergeSort(arr,0,n-1);
    for(int i=0;i<n;i++){
      std::cout << arr[i] << std::endl;
    }
    return 0;
}